waku/v2/node/storage/message/waku_message_store.nim

```diff
@@ -12,22 +13,49 @@ import
 
 export sqlite
 
+logScope:
+  topics = "wakuMessageStore"
+
 const TABLE_TITLE = "Message"
+const maxStoreOverflow = 1.3
 
 # The code in this file is an adaptation of the Sqlite KV Store found in nim-eth.
 # https://github.com/status-im/nim-eth/blob/master/eth/db/kvstore_sqlite3.nim
-#
-# Most of it is a direct copy, the only unique functions being `get` and `put`.
 
 type
+  # WakuMessageStore implements auto deletion as follows:
+  # The sqlite DB will store up to `storeCapacity` * `maxStoreOverflow` messages.
+  # If more messages are stored, the messages are sorted by `receiverTimestamp` and the oldest ones are deleted,
+  # bringing the total number of stored messages back to `storeCapacity*maxStoreOverflow/2`. The rationale for batch deleting is efficiency.
+  # We keep half of the overflow window because we delete the oldest messages with respect to `receiverTimestamp` instead of `senderTimestamp`.
+  # `ReceiverTimestamp` is guaranteed to be set, while senders could omit setting `senderTimestamp`.
+  # However, `receiverTimestamp` can differ from node to node for the same message.
+  # So sorting by `receiverTimestamp` might (slightly) prioritize some actually older messages and we compensate that by keeping half of the overflow window.
   WakuMessageStore* = ref object of MessageStore
     database*: SqliteDatabase
+    numMessages: int
+    storeCapacity: int # represents both the number of messages that are persisted in the sqlite DB (excl. the overflow window explained above), and the number of messages that get loaded via `getAll`.
+    storeMaxLoad: int  # = --store-capacity * maxStoreOverflow
  
-proc init*(T: type WakuMessageStore, db: SqliteDatabase): MessageStoreResult[T] =
+proc deleteOldest(db: WakuMessageStore): MessageStoreResult[void] =
+  let limit = int(float(db.storeCapacity) * maxStoreOverflow/2)
+  var deleteQuery = "DELETE FROM " & TABLE_TITLE & " " &
```

Comment:
It seems an expensive query that can take a lot of time to conclude (of course depends on the internals of the db, but in the worst case it can take `n`*`m` comparisons where `n` is the DB size and `m` is the selected portion), how about selecting the portion you want, then erasing the entire table and replacing its content with that already selected portion.

Reply:
We are indeed planning several DB and efficiency improvements for subsequent PRs.

Reply:
Yes. In the current implementation, most of the messages are kept and only a comparatively small part is deleted.
So, selecting the messages that are kept would take most of the time.
An exception would be if a very large DB of an older version is loaded.

My suggestion for making this query more efficient would be adding an index on `receiverTimestamp`,
which would reduce the cost for the select part to log(m).
I will open an issue for this (+ vacuum) as soon as this issue is closed.

While Index + vacuum will be addressed timely, future steps we envisioned are:

* pure SQL store
  - large storage capacity but slow
  - in-line with status-desktop requirement of keeping messages for 30 day
* adding an efficient cache

I could also add a store efficiency roadmap issue tracking these future goals.